framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -115,52 +126,42 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundleP
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
+  auto l = registration->Lock(); US_UNUSED(l);
```

Comment:
I agree, we shouldn't hold locks while calling client code. In this case, I think the only way a deadlock occurs is if the client's ```ServiceFactory::GetService``` implementation calls ```GetServiceReference/GetService``` on itself since the lock is on the ```ServiceRegistrationBasePrivate``` object associated with the ```ServiceFactory``` object. I'm sure a deadlock could occur, I don't know how likely that is to happen.

An alternative solution is to scope the locks such that ```GetServiceFromFactory(...)``` is unconditionally called and the return value is discarded if a cached service object already exists. This seems wasteful.

These are the only two options I could come up with that fixed the race condition. What are your thoughts?